GH-41816: [C++] Meson Build System Support

WillAyd opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

The Meson build system may be more user friendly to some developers, and may make it easier to perform tasks like valgrind, coverage, or ASAN/UBSAN coverage. There is also a prior art for using meson in the nanoarrow and arrow-adbc projects.

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

This PR implements a Meson configuration that can support the `ARROW_BUILD_TESTS` and `ARROW_COMPUTE` options, as well as the dependencies for those options

### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Not in CI. You can download and run `meson test` when built to run the test suite.

### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

No. This is strictly for developers

WillAyd commented:
@kou this is a simplified attempt at adding Meson to address your comment https://github.com/apache/arrow/issues/41816#issuecomment-2131511532

This is currently slower than the CMake configuration by a good deal (pending investigation) and the configuration file is not 100% complete, but this should give us an idea of what a Meson configuration may look like.

To use, from the cpp directory developers can:

```bash
meson setup builddir -Dtests=true -Dcompute=true
meson compile -C builddir
meson test -C builddir
```

For ASAN/UBSAN, users could simply:

```bash
meson setup builddir -Dtests=true -Dcompute=true -Db_sanitize=address,undefined
```

Or if the project is already setup run:

```bash
meson configure -C builddir -Db_sanitize=address,undefined
```

Coverage can be enabled with:

```bash
meson configure -C builddir -Db_coverage=true
```

and tests can be run under valgrind with:

```bash
meson test -C builddir --wrap='valgrind --track-origins=yes --leak-check=full' --print-errorlog
```

kou commented:
Thanks.

Can we start from more simplified version? For example, we don't need `libarrow_testing`, tests and so on at the first step.

We want to add a nightly CI for this to detect regression.

We want to update version information automatically in release process. For example: https://github.com/apache/arrow/blob/05569054edf4260eebbfbfd87ca524da05082afe/dev/release/utils-prepare.sh#L39-L44

(I can do it in this branch later.)

Anyway, I didn't know that `meson format` exists! I want to use it in `c_glib/` too.

kou commented:
Yes. Only minimal `libarrow` (e.g. no compute, no filesystem and so on) is the right starting point.

WillAyd commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,333 @@
+    objects: objlibs,
+    include_directories: [include_dir],
+    install: true,
+    # compute/expression.cc may have undefined IPC symbols in non-IPC builds
```
Wasn't sure if this was intentional or not in the existing code base

WillAyd commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,333 @@
+    override_options: {'b_lundef': 'false'},
+)
+
+# Meson does not allow you to glob for headers to install. See also
```
Meson is pretty strict about wildcards for the reasons outlined in their FAQ. Idiomatically, Meson would want you to put the files you want to install in a separate directory and call `install_subdir`, but that would go beyond the scope of this initial PR I think

WillAyd commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,333 @@
+arrow_so_version = (ver_major.to_int() * 100 + ver_minor.to_int()).to_string()
+arrow_full_so_version = '@0@.@1@.@2@'.format(arrow_so_version, ver_patch, 0)
+
+# TODO: The Meson generated .pc file does not include the Apache license
```
I did not research this in too much detail yet; figured I'd check if it was a big deal before investing time into a resolution

kou replied:
We don't need it.
`arrow.pc.in` has the license header because it's in this repository. Files that don't exist in this repository don't need it.

WillAyd commented on cpp/src/arrow/util/meson.build:
```diff
@@ -0,0 +1,103 @@
+conf_data.set('ARROW_SO_VERSION', arrow_so_version)
+conf_data.set('ARROW_FULL_SO_VERSION', arrow_full_so_version)
+
+# TODO: we may want to give Meson its own configuration file rather
```
Instead of reusing the `config.h.cmake` file we would create one with directives specifically for Meson. That might make things slightly easier on the Meson side, although I'm not sure it is worth it to have two different config files

kou replied:
We want to use `config.h.cmake` for both CMake and Meson (with `format: 'cmake@'`). It's for reducing maintenance cost.

WillAyd reviewed (commented):
Thanks @kou . Hope this is more in line with your expectation.

As far as nightlies go, can you point me to an existing nightly CI setup in the repo? I was able to grep for some R nightlies, but not sure if there is existing infrastructure for C++ nightly jobs where this would be better placed

kou commented on cpp/src/arrow/util/meson.build:
```diff
@@ -0,0 +1,103 @@
+    configuration: conf_data,
+    format: 'cmake@',
+    install: true,
+    install_dir: 'arrow',
```
```suggestion
    install_dir: 'arrow/util',
```

kou commented on cpp/src/arrow/util/meson.build:
```diff
@@ -0,0 +1,103 @@
+conf_data.set('ARROW_PACKAGE_KIND', get_option('package_kind'))
+
+foreach cmakedefine : [
+    'ARROW_COMPUTE',
+    'ARROW_CSV',
+    'ARROW_CUDA',
+    'ARROW_DATASET',
+    'ARROW_FILESYSTEM',
+    'ARROW_FLIGHT',
+    'ARROW_FLIGHT_SQL',
+    'ARROW_IPC',
+    'ARROW_JEMALLOC',
+    'ARROW_JEMALLOC_VENDORED',
+    'ARROW_JSON',
+    'ARROW_MIMALLOC',
+    'ARROW_ORC',
+    'ARROW_PARQUET',
+    'ARROW_SUBSTRAIT',
+    'ARROW_AZURE',
+    'ARROW_ENABLE_THREADING',
+    'ARROW_GCS',
+    'ARROW_HDFS',
+    'ARROW_S3',
+    'ARROW_USE_GLOG',
+    'ARROW_USE_NATIVE_INT128',
+    'ARROW_WITH_BROTLI',
+    'ARROW_WITH_BZ2',
+    'ARROW_WITH_LZ4',
+    'ARROW_WITH_MUSL',
+    'ARROW_WITH_OPENTELEMETRY',
+    'ARROW_WITH_RE2',
+    'ARROW_WITH_SNAPPY',
+    'ARROW_WITH_UCX',
+    'ARROW_WITH_UTF8PROC',
+    'ARROW_WITH_ZLIB',
+    'ARROW_WITH_ZSTD',
+    'PARQUET_REQUIRE_ENCRYPTION',
+]
+    conf_data.set(cmakedefine, false)
+endforeach
```
It seems that we don't need to use `foreach` here:

```meson
conf_data.set('ARROW_COMPUTE', false)
conf_data.set('ARROW_CSV', false)
...

kou commented on cpp/src/arrow/util/meson.build:
```diff
@@ -0,0 +1,103 @@
+    output: 'config_internal.h',
+    configuration: internal_conf_data,
+    format: 'cmake@',
+    install: true,
+    install_dir: 'arrow',
```
We don't need to install `_internal.h`:

```suggestion
```

kou reviewed (commented):
Can I push some commits to this branch for nightly CI and auto version update?

WillAyd commented:
Sure no problem

WillAyd commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,333 @@
+
+# Meson does not natively support object libraries
+# https://github.com/mesonbuild/meson/issues/13843
+objlib_sources = {
```
I have a pretty poor understanding of how object libraries actually work, but I don't think we can just convert these to using the `library` call unless we want the linker to allow undefined symbols in these libraries as well (?)

Trying to do what I think you suggest generates a huge amount of errors like:

```
[7/96] Linking target src/arrow/libarrow_io.so
FAILED: src/arrow/libarrow_io.so 
c++  -o src/arrow/libarrow_io.so src/arrow/libarrow_io.so.p/io_buffered.cc.o src/arrow/libarrow_io.so.p/io_caching.cc.o src/arrow/libarrow_io.so.p/io_compressed.cc.o src/arrow/libarrow_io.so.p/io_file.cc.o src/arrow/libarrow_io.so.p/io_hdfs.cc.o src/arrow/libarrow_io.so.p/io_hdfs_internal.cc.o src/arrow/libarrow_io.so.p/io_interfaces.cc.o src/arrow/libarrow_io.so.p/io_memory.cc.o src/arrow/libarrow_io.so.p/io_slow.cc.o src/arrow/libarrow_io.so.p/io_stdio.cc.o src/arrow/libarrow_io.so.p/io_transform.cc.o -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 -shared -fPIC -Wl,--start-group -Wl,-soname,libarrow_io.so -Wl,--end-group
/usr/bin/ld: src/arrow/libarrow_io.so.p/io_buffered.cc.o: in function `arrow::io::BufferedInputStream::SetBufferSize(long)':
buffered.cc:(.text+0x13a9): undefined reference to `arrow::Buffer::CheckCPU() const'
/usr/bin/ld: buffered.cc:(.text+0x13b1): undefined reference to `arrow::Buffer::CheckMutable() const'
/usr/bin/ld: buffered.cc:(.text+0x1407): undefined reference to `arrow::util::detail::StringStreamWrapper::StringStreamWrapper()'
/usr/bin/ld: buffered.cc:(.text+0x149e): undefined reference to `arrow::util::detail::StringStreamWrapper::str[abi:cxx11]()'
/usr/bin/ld: buffered.cc:(.text+0x14a6): undefined reference to `arrow::util::detail::StringStreamWrapper::~StringStreamWrapper()'
/usr/bin/ld: buffered.cc:(.text+0x14b6): undefined reference to `arrow::Status::Status(arrow::StatusCode, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```

kou replied:
Ah, we can remove `objlib_sources` and `static_library()` for them entirely.
We can just add sources in `objlib_sources` to `arrow_srcs` and use it in one `library()` (that already exists in this PR).

kou replied:
(We don't need objlib feature entirely with Meson.)

WillAyd replied:
Makes sense. I've still kept the top level dict for parity with CMake, and since some of the sources vary depending on compilation options (like arrow compute). Let me know if this is more in line with what you are thinking

kou commented:
@github-actions crossbow submit test-conda-cpp-meson

kou commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,325 @@
+dl_dep = dependency('dl')
+threads_dep = dependency('threads')
+
+# Meson does not natively support object libraries
+# https://github.com/mesonbuild/meson/issues/13843
```
Can we remove this comment?
We don't need this comment now, right?

kou commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,325 @@
+        'device.h',
+        'extension_type.h',
+        'memory_pool.h',
+        'memory_pool_test.h',
```
```suggestion
```

kou commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,325 @@
+        'extension_type.h',
+        'memory_pool.h',
+        'memory_pool_test.h',
+        'pch.h',
```
```suggestion
```

kou commented on cpp/src/arrow/meson.build:
```diff
@@ -0,0 +1,325 @@
+        'scalar.h',
+        'sparse_tensor.h',
+        'status.h',
+        'stl_allocator.h',
+        'stl.h',
+        'stl_iterator.h',
```
Could you sort them?

raulcd commented on ci/scripts/cpp_test.sh:
```diff
@@ -89,14 +89,21 @@ pushd ${build_dir}
-ctest \
+if [ "${ARROW_USE_MESON:-OFF}" = "ON" ]; then
+  ARROW_BUILD_EXAMPLES=OFF # TODO: Remove this
+  meson test \
+    --print-errorlogs \
+    "$@"
```
We aren't building any tests now, are we? On the current CI:
```
+ meson test --print-errorlogs
No tests defined.
```

WillAyd replied:
For now we aren't actually building any tests - just creating an MRE for Meson with libarrow

kou replied:
Right. We'll add tests later.

WillAyd commented:
> I've rebased on main, added support for updating version in release process and added nightly CI job for Meson.
> 
> Could you update the PR description?

I've addressed your feedback. Not sure I am doing what you are asking with respect to the PR description but I changed it to be more exact. Let me know if there is something else should be doing.

kou commented:
> Not sure I am doing what you are asking with respect to the PR description but I changed it to be more exact.

We use the PR description for a commit message when we merge this. So we want to update the PR description to reflect the latest PR changes.

BTW, it seems that you didn't change the PR description yet. (You might forget to save your PR description change.)

kou commented:
@github-actions crossbow submit test-conda-cpp-meson

WillAyd commented:
Do you mean the commit message? I can squash up and change that if so. On Github I've already change the PR title

kou commented:
I mean the merge commit's commit message not commit messages of commits in this PRs.

You don't need to squash commits in this PR. Because we use "Squash and merge" feature provided by GitHub. We use the following commit message:

```text
${PR_TITLE}

${PR_DESCRIPTION}
```

For example, https://github.com/apache/arrow/commit/ec3d2839d102cc44d368c154cfa756eb946909ee 's commit message uses https://github.com/apache/arrow/pull/44375 's title and description. It doesn't use commit messages of commits in https://github.com/apache/arrow/pull/44375 .

> On Github I've already change the PR title

Could you also update the PR description?

WillAyd commented:
Ah got it - I've updated the OP

kou reviewed (approved):
+1
